File: Python/0015-m-3sum.py
"""
Sort and use two pointers
Time: O(n^2)
Space: O(1)
"""
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        for i in range(len(nums) - 2):
            # remove duplicate 2sum target if nums[i] == nums[i - 1]
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            left = i + 1
            right = len(nums) - 1
            while left < right:
                s = nums[i] + nums[left] + nums[right]
                if s < 0:
                    left += 1
                elif s > 0: 
                    right -= 1
                else:
                    res.append([nums[i], nums[left], nums[right]])
                    left += 1
                    right -= 1
                    # remove duplicate
                    while left < right and nums[left] == nums[left - 1]:
                        left += 1
                    while left < right and nums[right] == nums[right + 1]:
                        right -= 1
        return res

File: leetcode/C++/0015-m-3sum.cpp
// Two pointers
// Time: O(N^2)
// Space: O(1)
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        if (nums.size() < 3) return res;
        sort(nums.begin(), nums.end());
        for (auto i = nums.begin(); i < nums.end() - 2; i++) {
            auto j = i + 1, k = nums.end() - 1;
            if (i > nums.begin() && *i == *(i - 1))
                continue;
            while (j < k) {
                if (*i + *j + *k < 0) {
                    j++;
                    while (*j == *(j - 1) && j < k)
                        j++;
                } else if (*i + *j + *k > 0) {
                    k--;
                    while (*k == *(k + 1) && j < k)
                        k--;
                } else {
                    res.push_back({*i, *j, *k});
                    j++;
                    k--;
                    while (*j == *(j - 1) && *k == *(k + 1) && j < k)
                        j++;
                }
            }
        }
        return res;
    }
};

// HashSet + TwoSum
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        for (int i = 0; i < nums.size() && nums[i] <= 0; ++i) {
            if (i == 0 || nums[i-1] != nums[i])
                twoSum(nums, i, res);
        }
        return res;
    }
    
    void twoSum(vector<int>& nums, int i, vector<vector<int>>& res) {
        unordered_set<int> seen;
        for (int j = i + 1; j < nums.size(); j++) {
            int complement = -nums[i] - nums[j];
            if (seen.find(complement) != seen.end()) {
                res.push_back({nums[i], nums[j], complement});
                while (j < nums.size() - 1 && nums[j] == nums[j + 1])
                    j++;
            }
            seen.insert(nums[j]);
        }
    }
};